File: assignment-1/swift/Assignment1.swift
class User {
    private var name: String? // allows nil

    init(_ name: String?) {
        self.name = name
    }

    func equals(_ obj: Any?) -> Bool {
        guard let other = obj as? User else { return false }
        return self.name == other.name
    }
}

class SpecialUser: User {
    private var id: Int

    init(_ name: String?, _ id: Int) {
        self.id = id
        super.init(name)
    }

    override func equals(_ obj: Any?) -> Bool {
        guard let other = obj as? SpecialUser else { return false }
        return super.equals(obj) && other.id == self.id
    }
}

// Main
var u: User
var v: User

u = User("Student1")
print(u.equals(u))
print()

u = User("Student1")
print(u.equals(nil))
print()

u = User("Student1")
print(u.equals("Student1"))
print()

u = User("Student1")
v = SpecialUser("Student1", 5)
print(u.equals(v))
print(v.equals(u))
print()

u = User("Student1")
v = User(nil)
print(v.equals(u))
print()

u = User("Student1")
v = User(nil)
print(u.equals(v))
print()

File: assignment-1/rust/src/assignment1.rs
// Assignment1.rs

use crate::user::User;

pub fn run() {
    println!("Running Assignment 1 in Rust");
    println!();

    // u = new User("Student1");
    // System.out.println(u.equals(u));
    let u = User::new_regular(Some("Student1"));
    println!("{}", u == u);
    println!();

    // In Java, u.equals(null) is false.
    // In Rust, we can't compare a User with None directly.
    println!("// In Java, u.equals(null) is false. In Rust, you can't compare a User with None directly.");
    println!();

    // In Java, u.equals("Student1") is false.
    // In Rust, you can't compare a User with a &str directly.
    println!("// In Java, u.equals(\"Student1\") is false. In Rust, you can't compare a User with a &str directly.");
    println!();

    // u = new User("Student1");
    // v = new SpecialUser("Student1", 5);
    // System.out.println(u.equals(v));
    // System.out.println(v.equals(u));
    let u = User::new_regular(Some("Student1"));
    let v = User::new_special(Some("Student1"), 5);
    println!("{}", u == v); // Regular == Special
    println!("{}", v == u); // Special == Regular
    println!();

    // u = new User("Student1");
    // v = new User(null);
    // System.out.println(v.equals(u));
    let u = User::new_regular(Some("Student1"));
    let v = User::new_regular(None);
    println!("{}", v == u);
    println!();

    // u = new User("Student1");
    // v = new User(null);
    // System.out.println(u.equals(v));
    let u = User::new_regular(Some("Student1"));
    let v = User::new_regular(None);
    println!("{}", u == v);
    println!();
}
